File: backend/app/api/gl.py
"""
ARIA ERP - General Ledger API
Provides Chart of Accounts, Journal Entries, and Financial Reports
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime, date
from pydantic import BaseModel, Field

from core.database import get_db
from core.auth import get_current_user
from models.accounting import (
    ChartOfAccounts, GeneralLedger, GeneralLedgerLine,
    AccountType, AccountSubType, JournalStatus, JournalEntryType
)
from models.user import User

router = APIRouter(prefix="/api/erp/gl", tags=["General Ledger"])


# Pydantic Models
class AccountCreate(BaseModel):
    account_code: str = Field(..., description="Account code (e.g., 1000, 4100)")
    account_name: str = Field(..., description="Account name")
    description: Optional[str] = None
    account_type: AccountType
    account_subtype: AccountSubType
    parent_account_id: Optional[int] = None
    is_control_account: bool = False
    accepts_posting: bool = True
    is_tax_account: bool = False
    vat_rate: float = 0.0


class AccountUpdate(BaseModel):
    account_name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    vat_rate: Optional[float] = None


class AccountResponse(BaseModel):
    id: int
    account_code: str
    account_name: str
    description: Optional[str]
    account_type: str
    account_subtype: str
    parent_account_id: Optional[int]
    is_control_account: bool
    is_system_account: bool
    accepts_posting: bool
    is_tax_account: bool
    vat_rate: float
    debit_balance: float
    credit_balance: float
    current_balance: float
    is_active: bool
    created_at: datetime
    
    class Config:
        from_attributes = True


class JournalLineCreate(BaseModel):
    account_id: int
    debit: float = 0.0
    credit: float = 0.0
    description: Optional[str] = None
    department: Optional[str] = None
    project: Optional[str] = None
    cost_center: Optional[str] = None


class JournalEntryCreate(BaseModel):
    entry_type: JournalEntryType
    entry_date: date
    description: str
    reference: Optional[str] = None
    notes: Optional[str] = None
    lines: List[JournalLineCreate]


class JournalEntryResponse(BaseModel):
    id: int
    journal_number: str
    entry_type: str
    entry_date: date
    posting_date: Optional[datetime]
    period: str
    financial_year: int
    description: str
    reference: Optional[str]
    status: str
    total_debit: float
    total_credit: float
    created_at: datetime
    
    class Config:
        from_attributes = True


@router.get("/accounts", response_model=List[AccountResponse])
async def list_accounts(
    company_id: str = Query(..., description="Company ID"),
    account_type: Optional[AccountType] = None,
    is_active: bool = True,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all chart of accounts for a company"""
    query = db.query(ChartOfAccounts).filter(
        ChartOfAccounts.tenant_id == company_id,
        ChartOfAccounts.is_active == is_active
    )
    
    if account_type:
        query = query.filter(ChartOfAccounts.account_type == account_type)
    
    accounts = query.order_by(ChartOfAccounts.account_code).all()
    return accounts


@router.get("/accounts/{account_id}", response_model=AccountResponse)
async def get_account(
    account_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific account by ID"""
    account = db.query(ChartOfAccounts).filter(ChartOfAccounts.id == account_id).first()
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    return account


@router.post("/accounts", response_model=AccountResponse)
async def create_account(
    account: AccountCreate,
    company_id: str = Query(..., description="Company ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new chart of accounts entry"""
    existing = db.query(ChartOfAccounts).filter(
        ChartOfAccounts.tenant_id == company_id,
        ChartOfAccounts.account_code == account.account_code
    ).first()
    
    if existing:
        raise HTTPException(status_code=400, detail="Account code already exists")
    
    db_account = ChartOfAccounts(
        tenant_id=company_id,
        account_code=account.account_code,
        account_name=account.account_name,
        description=account.description,
        account_type=account.account_type,
        account_subtype=account.account_subtype,
        parent_account_id=account.parent_account_id,
        is_control_account=account.is_control_account,
        accepts_posting=account.accepts_posting,
        is_tax_account=account.is_tax_account,
        vat_rate=account.vat_rate,
        created_by=current_user.email if current_user else None
    )
    
    db.add(db_account)
    db.commit()
    db.refresh(db_account)
    
    return db_account


@router.put("/accounts/{account_id}", response_model=AccountResponse)
async def update_account(
    account_id: int,
    account: AccountUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update an existing account"""
    db_account = db.query(ChartOfAccounts).filter(ChartOfAccounts.id == account_id).first()
    if not db_account:
        raise HTTPException(status_code=404, detail="Account not found")
    
    if db_account.is_system_account:
        raise HTTPException(status_code=400, detail="Cannot modify system accounts")
    
    # Update fields
    if account.account_name is not None:
        db_account.account_name = account.account_name
    if account.description is not None:
        db_account.description = account.description
    if account.is_active is not None:
        db_account.is_active = account.is_active
    if account.vat_rate is not None:
        db_account.vat_rate = account.vat_rate
    
    db_account.updated_at = datetime.utcnow()
    
    db.commit()
    db.refresh(db_account)
    
    return db_account


@router.delete("/accounts/{account_id}")
async def delete_account(
    account_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete an account (soft delete by setting is_active=False)"""
    db_account = db.query(ChartOfAccounts).filter(ChartOfAccounts.id == account_id).first()
    if not db_account:
        raise HTTPException(status_code=404, detail="Account not found")
    
    if db_account.is_system_account:
        raise HTTPException(status_code=400, detail="Cannot delete system accounts")
    
    has_transactions = db.query(GeneralLedgerLine).filter(
        GeneralLedgerLine.account_id == account_id
    ).first()
    
    if has_transactions:
        db_account.is_active = False
        db_account.updated_at = datetime.utcnow()
        db.commit()
        return {"message": "Account deactivated (has transactions)"}
    else:
        db.delete(db_account)
        db.commit()
        return {"message": "Account deleted"}


@router.get("/journal-entries", response_model=List[JournalEntryResponse])
async def list_journal_entries(
    company_id: str = Query(..., description="Company ID"),
    status: Optional[JournalStatus] = None,
    entry_type: Optional[JournalEntryType] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List journal entries for a company"""
    query = db.query(GeneralLedger).filter(GeneralLedger.tenant_id == company_id)
    
    if status:
        query = query.filter(GeneralLedger.status == status)
    if entry_type:
        query = query.filter(GeneralLedger.entry_type == entry_type)
    if start_date:
        query = query.filter(GeneralLedger.entry_date >= start_date)
    if end_date:
        query = query.filter(GeneralLedger.entry_date <= end_date)
    
    entries = query.order_by(GeneralLedger.entry_date.desc()).limit(limit).all()
    return entries


@router.get("/journal-entries/{entry_id}", response_model=JournalEntryResponse)
async def get_journal_entry(
    entry_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific journal entry by ID"""
    entry = db.query(GeneralLedger).filter(GeneralLedger.id == entry_id).first()
    if not entry:
        raise HTTPException(status_code=404, detail="Journal entry not found")
    return entry


@router.post("/journal-entries", response_model=JournalEntryResponse)
async def create_journal_entry(
    entry: JournalEntryCreate,
    company_id: str = Query(..., description="Company ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new journal entry"""
    total_debit = sum(line.debit for line in entry.lines)
    total_credit = sum(line.credit for line in entry.lines)
    
    if abs(total_debit - total_credit) > 0.01:
        raise HTTPException(
            status_code=400,
            detail=f"Journal entry not balanced: DR={total_debit}, CR={total_credit}"
        )
    
    year = entry.entry_date.year
    period = entry.entry_date.strftime("%Y-%m")
    
    last_entry = db.query(GeneralLedger).filter(
        GeneralLedger.tenant_id == company_id,
        GeneralLedger.period == period
    ).order_by(GeneralLedger.id.desc()).first()
    
    seq = 1 if not last_entry else int(last_entry.journal_number.split("-")[-1]) + 1
    journal_number = f"JE-{year}-{seq:05d}"
    
    # Create journal entry
    db_entry = GeneralLedger(
        tenant_id=company_id,
        journal_number=journal_number,
        entry_type=entry.entry_type,
        entry_date=entry.entry_date,
        period=period,
        financial_year=year,
        description=entry.description,
        reference=entry.reference,
        notes=entry.notes,
        status=JournalStatus.DRAFT,
        total_debit=total_debit,
        total_credit=total_credit,
        created_by=current_user.email if current_user else None
    )
    
    db.add(db_entry)
    db.flush()  # Get the ID
    
    for idx, line in enumerate(entry.lines, 1):
        db_line = GeneralLedgerLine(
            tenant_id=company_id,
            journal_id=db_entry.id,
            line_number=idx,
            account_id=line.account_id,
            debit=line.debit,
            credit=line.credit,
            description=line.description or entry.description,
            department=line.department,
            project=line.project,
            cost_center=line.cost_center
        )
        db.add(db_line)
    
    db.commit()
    db.refresh(db_entry)
    
    return db_entry


@router.post("/journal-entries/{entry_id}/post")
async def post_journal_entry(
    entry_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Post a journal entry to the general ledger"""
    entry = db.query(GeneralLedger).filter(GeneralLedger.id == entry_id).first()
    if not entry:
        raise HTTPException(status_code=404, detail="Journal entry not found")
    
    if entry.status != JournalStatus.DRAFT:
        raise HTTPException(status_code=400, detail="Only draft entries can be posted")
    
    if not entry.is_balanced():
        raise HTTPException(status_code=400, detail="Journal entry is not balanced")
    
    entry.status = JournalStatus.POSTED
    entry.posting_date = datetime.utcnow()
    entry.posted_by = current_user.email if current_user else None
    
    lines = db.query(GeneralLedgerLine).filter(GeneralLedgerLine.journal_id == entry_id).all()
    for line in lines:
        account = db.query(ChartOfAccounts).filter(ChartOfAccounts.id == line.account_id).first()
        if account:
            account.debit_balance += line.debit
            account.credit_balance += line.credit
            
            if account.account_type in [AccountType.ASSET, AccountType.EXPENSE]:
                account.current_balance = account.debit_balance - account.credit_balance
            else:
                account.current_balance = account.credit_balance - account.debit_balance
    
    db.commit()
    
    return {"message": "Journal entry posted successfully", "journal_number": entry.journal_number}


@router.delete("/journal-entries/{entry_id}")
async def delete_journal_entry(
    entry_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete a draft journal entry"""
    entry = db.query(GeneralLedger).filter(GeneralLedger.id == entry_id).first()
    if not entry:
        raise HTTPException(status_code=404, detail="Journal entry not found")
    
    if entry.status != JournalStatus.DRAFT:
        raise HTTPException(status_code=400, detail="Only draft entries can be deleted")
    
    db.query(GeneralLedgerLine).filter(GeneralLedgerLine.journal_id == entry_id).delete()
    
    db.delete(entry)
    db.commit()
    
    return {"message": "Journal entry deleted"}


@router.get("/reports/trial-balance")
async def get_trial_balance(
    company_id: str = Query(..., description="Company ID"),
    as_of_date: Optional[date] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Generate trial balance report"""
    accounts = db.query(ChartOfAccounts).filter(
        ChartOfAccounts.tenant_id == company_id,
        ChartOfAccounts.is_active == True
    ).order_by(ChartOfAccounts.account_code).all()
    
    trial_balance = []
    total_debits = 0.0
    total_credits = 0.0
    
    for account in accounts:
        if account.current_balance != 0:
            if account.account_type in [AccountType.ASSET, AccountType.EXPENSE]:
                debit = abs(account.current_balance) if account.current_balance > 0 else 0
                credit = abs(account.current_balance) if account.current_balance < 0 else 0
            else:
                credit = abs(account.current_balance) if account.current_balance > 0 else 0
                debit = abs(account.current_balance) if account.current_balance < 0 else 0
            
            trial_balance.append({
                "account_code": account.account_code,
                "account_name": account.account_name,
                "account_type": account.account_type.value,
                "debit": debit,
                "credit": credit
            })
            
            total_debits += debit
            total_credits += credit
    
    return {
        "company_id": company_id,
        "as_of_date": str(as_of_date) if as_of_date else str(date.today()),
        "accounts": trial_balance,
        "total_debits": total_debits,
        "total_credits": total_credits,
        "balanced": abs(total_debits - total_credits) < 0.01,
        "difference": abs(total_debits - total_credits)
    }


@router.get("/reports/balance-sheet")
async def get_balance_sheet(
    company_id: str = Query(..., description="Company ID"),
    as_of_date: Optional[date] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Generate balance sheet report"""
    accounts = db.query(ChartOfAccounts).filter(
        ChartOfAccounts.tenant_id == company_id,
        ChartOfAccounts.is_active == True
    ).all()
    
    assets = []
    liabilities = []
    equity = []
    
    total_assets = 0.0
    total_liabilities = 0.0
    total_equity = 0.0
    
    for account in accounts:
        if account.current_balance != 0:
            balance = abs(account.current_balance)
            
            if account.account_type == AccountType.ASSET:
                assets.append({
                    "account_code": account.account_code,
                    "account_name": account.account_name,
                    "account_subtype": account.account_subtype.value,
                    "amount": balance
                })
                total_assets += balance
            elif account.account_type == AccountType.LIABILITY:
                liabilities.append({
                    "account_code": account.account_code,
                    "account_name": account.account_name,
                    "account_subtype": account.account_subtype.value,
                    "amount": balance
                })
                total_liabilities += balance
            elif account.account_type == AccountType.EQUITY:
                equity.append({
                    "account_code": account.account_code,
                    "account_name": account.account_name,
                    "account_subtype": account.account_subtype.value,
                    "amount": balance
                })
                total_equity += balance
    
    return {
        "company_id": company_id,
        "as_of_date": str(as_of_date) if as_of_date else str(date.today()),
        "assets": {
            "accounts": assets,
            "total": total_assets
        },
        "liabilities": {
            "accounts": liabilities,
            "total": total_liabilities
        },
        "equity": {
            "accounts": equity,
            "total": total_equity
        },
        "total_liabilities_equity": total_liabilities + total_equity,
        "balanced": abs(total_assets - (total_liabilities + total_equity)) < 0.01,
        "difference": total_assets - (total_liabilities + total_equity)
    }

File: workers-api/src/routes/gl.ts
/**
 * General Ledger Routes
 * Chart of Accounts, Journal Entries, Trial Balance
 */

import { Hono } from 'hono';
import { getSecureCompanyId } from '../middleware/auth';

interface Env {
  DB: D1Database;
  JWT_SECRET: string;
}

const app = new Hono<{ Bindings: Env }>();

// ==================== CHART OF ACCOUNTS ====================

// List all accounts
app.get('/chart-of-accounts', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const search = c.req.query('search') || '';
    const type = c.req.query('type') || '';
    
    let query = 'SELECT * FROM gl_accounts WHERE company_id = ?';
    const params: any[] = [companyId];
    
    if (search) {
      query += ' AND (account_code LIKE ? OR account_name LIKE ?)';
      params.push(`%${search}%`, `%${search}%`);
    }
    
    if (type) {
      query += ' AND account_type = ?';
      params.push(type);
    }
    
    query += ' ORDER BY account_code';
    
    const result = await c.env.DB.prepare(query).bind(...params).all();
    
    return c.json({
      accounts: result.results || [],
      total: result.results?.length || 0
    });
  } catch (error) {
    console.error('Error loading accounts:', error);
    return c.json({ error: 'Failed to load accounts' }, 500);
  }
});

// Create account
app.post('/chart-of-accounts', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const body = await c.req.json();
    const { account_code, account_name, account_type, account_category, is_active } = body;
    
    if (!account_code || !account_name || !account_type) {
      return c.json({ error: 'Account code, name, and type are required' }, 400);
    }
    
    // Check for duplicate account code
    const existing = await c.env.DB.prepare(
      'SELECT id FROM gl_accounts WHERE company_id = ? AND account_code = ?'
    ).bind(companyId, account_code).first();
    
    if (existing) {
      return c.json({ error: 'Account code already exists' }, 409);
    }
    
    const id = crypto.randomUUID();
    const now = new Date().toISOString();
    
    await c.env.DB.prepare(`
      INSERT INTO gl_accounts (id, company_id, account_code, account_name, account_type, account_category, is_active, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).bind(id, companyId, account_code, account_name, account_type, account_category || null, is_active !== false ? 1 : 0, now, now).run();
    
    return c.json({
      id,
      account_code,
      account_name,
      account_type,
      account_category,
      is_active: is_active !== false,
      created_at: now
    }, 201);
  } catch (error) {
    console.error('Error creating account:', error);
    return c.json({ error: 'Failed to create account' }, 500);
  }
});

// Update account
app.put('/chart-of-accounts/:code', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const code = c.req.param('code');
    const body = await c.req.json();
    const { account_name, account_type, account_category, is_active } = body;
    
    const existing = await c.env.DB.prepare(
      'SELECT id FROM gl_accounts WHERE company_id = ? AND account_code = ?'
    ).bind(companyId, code).first();
    
    if (!existing) {
      return c.json({ error: 'Account not found' }, 404);
    }
    
    const now = new Date().toISOString();
    
    await c.env.DB.prepare(`
      UPDATE gl_accounts 
      SET account_name = ?, account_type = ?, account_category = ?, is_active = ?, updated_at = ?
      WHERE company_id = ? AND account_code = ?
    `).bind(account_name, account_type, account_category || null, is_active !== false ? 1 : 0, now, companyId, code).run();
    
    return c.json({ message: 'Account updated successfully' });
  } catch (error) {
    console.error('Error updating account:', error);
    return c.json({ error: 'Failed to update account' }, 500);
  }
});

// Delete account
app.delete('/chart-of-accounts/:code', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const code = c.req.param('code');
    
    // Check if account has journal entries
    const hasEntries = await c.env.DB.prepare(`
      SELECT COUNT(*) as count FROM journal_entry_lines jel
      JOIN journal_entries je ON jel.journal_entry_id = je.id
      JOIN gl_accounts ga ON jel.account_id = ga.id
      WHERE ga.company_id = ? AND ga.account_code = ?
    `).bind(companyId, code).first();
    
    if (hasEntries && (hasEntries as any).count > 0) {
      return c.json({ error: 'Cannot delete account with journal entries' }, 400);
    }
    
    await c.env.DB.prepare(
      'DELETE FROM gl_accounts WHERE company_id = ? AND account_code = ?'
    ).bind(companyId, code).run();
    
    return c.json({ message: 'Account deleted successfully' });
  } catch (error) {
    console.error('Error deleting account:', error);
    return c.json({ error: 'Failed to delete account' }, 500);
  }
});

// ==================== JOURNAL ENTRIES ====================

// List journal entries
app.get('/journal-entries', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const result = await c.env.DB.prepare(`
      SELECT je.*, 
        (SELECT GROUP_CONCAT(jel.account_id || ':' || jel.debit_amount || ':' || jel.credit_amount, '|')
         FROM journal_entry_lines jel WHERE jel.journal_entry_id = je.id) as lines_data
      FROM journal_entries je
      WHERE je.company_id = ?
      ORDER BY je.entry_date DESC, je.entry_number DESC
    `).bind(companyId).all();
    
    const entries = (result.results || []).map((entry: any) => ({
      ...entry,
      lines: entry.lines_data ? entry.lines_data.split('|').map((line: string, idx: number) => {
        const [account_id, debit, credit] = line.split(':');
        return {
          line_number: idx + 1,
          account_id,
          debit_amount: parseFloat(debit) || 0,
          credit_amount: parseFloat(credit) || 0
        };
      }) : []
    }));
    
    return c.json({ entries });
  } catch (error) {
    console.error('Error loading journal entries:', error);
    return c.json({ error: 'Failed to load journal entries' }, 500);
  }
});

// Create journal entry
app.post('/journal-entries', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const body = await c.req.json();
    const { entry_date, description, reference, lines } = body;
    
    if (!entry_date || !description || !lines || lines.length < 2) {
      return c.json({ error: 'Entry date, description, and at least 2 lines are required' }, 400);
    }
    
    // Validate balanced entry
    const totalDebits = lines.reduce((sum: number, line: any) => sum + (parseFloat(line.debit_amount) || 0), 0);
    const totalCredits = lines.reduce((sum: number, line: any) => sum + (parseFloat(line.credit_amount) || 0), 0);
    
    if (Math.abs(totalDebits - totalCredits) > 0.01) {
      return c.json({ error: 'Journal entry must be balanced (debits must equal credits)' }, 400);
    }
    
    // Generate entry number
    const lastEntry = await c.env.DB.prepare(
      "SELECT entry_number FROM journal_entries WHERE company_id = ? ORDER BY entry_number DESC LIMIT 1"
    ).bind(companyId).first();
    
    let entryNumber = 'JE-0001';
    if (lastEntry && (lastEntry as any).entry_number) {
      const lastNum = parseInt((lastEntry as any).entry_number.replace('JE-', '')) || 0;
      entryNumber = `JE-${String(lastNum + 1).padStart(4, '0')}`;
    }
    
    const id = crypto.randomUUID();
    const now = new Date().toISOString();
    
    await c.env.DB.prepare(`
      INSERT INTO journal_entries(id, company_id, entry_number, entry_date, description, reference, status, total_debit, total_credit, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).bind(id, companyId, entryNumber, entry_date, description, reference || null, 'draft', totalDebits, totalCredits, now, now).run();
    
    // Insert journal entry lines
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      if (!line.account_code || (line.debit_amount === 0 && line.credit_amount === 0)) continue;
      
      // Get account ID from account code
      const account = await c.env.DB.prepare(
        'SELECT id FROM gl_accounts WHERE company_id = ? AND account_code = ?'
      ).bind(companyId, line.account_code).first();
      
      if (!account) {
        return c.json({ error: `Account ${line.account_code} not found` }, 400);
      }
      
      const lineId = crypto.randomUUID();
      await c.env.DB.prepare(`
        INSERT INTO journal_entry_lines (id, journal_entry_id, line_number, account_id, description, debit_amount, credit_amount)
        VALUES (?, ?, ?, ?, ?, ?, ?)
      `).bind(lineId, id, i + 1, (account as any).id, line.description || null, line.debit_amount || 0, line.credit_amount || 0).run();
    }
    
    return c.json({
      id,
      entry_number: entryNumber,
      entry_date,
      description,
      status: 'draft',
      total_debit: totalDebits,
      total_credit: totalCredits
    }, 201);
  } catch (error) {
    console.error('Error creating journal entry:', error);
    return c.json({ error: 'Failed to create journal entry' }, 500);
  }
});

// Post journal entry
app.post('/journal-entries/:id/post', async (c) => {
  const companyId = await getSecureCompanyId(c);

  try {
    const id = c.req.param('id');
    
    const entry = await c.env.DB.prepare(
      'SELECT * FROM journal_entries WHERE id = ? AND company_id = ?'
    ).bind(id, companyId).first();
    
    if (!entry) {
      return c.json({ error: 'Journal entry not found' }, 404);
    }
    
    if ((entry as any).status === 'posted') {
      return c.json({ error: 'Journal entry is already posted' }, 400);
    }
    
    const now = new Date().toISOString();
    
    await c.env.DB.prepare(`
      UPDATE journal_entries SET status = 'posted', posted_at = ?, updated_at = ? WHERE id = ?
    `).bind(now, now, id).run();
    
    return c.json({ message: 'Journal entry posted successfully' });
  } catch (error) {
    console.error('Error posting journal entry:', error);
    return c.json({ error: 'Failed to post journal entry' }, 500);
  }
});

export default app;
